Add unit tests for Source Context

Deren-Liao opened the pull request:
#988 
Also update the version number to 1.0.0 from 1.0.0.alpha3 so as to make it GA ready.

jskeet commented:
I'm on vacation all week. @evildour may be able to review.

Deren-Liao commented:
Hi Chris,

May you help figure out the build error?  The build passed in my Visual Studio as well as dotnet build command.

jskeet commented:
Chris is on holiday as well this week. But it looks like you've modified the project file directly without rerunning `generateprojects.sh` from the root directory. You might want to just delete the project file and rerun that - it'll create a project file which will probably be what you want it to be...

evildour commented:
Yes, I'll take a look now and help with any other build issues that might be occurring after running generateprojects.sh

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,80 @@
+            Assert.Equal(sourceContext.ContextCase, SourceContext.ContextOneofCase.None);
+        }
+
+        [Fact] void DefaultEmptySourceContext()
```
Nit: keep the `[Fact]` attribute on its own line and make this method public. Same goes for other tests below.

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,80 @@
+
+        [Fact] void ReadIOError()
+        {
+            SourceContext.s_fileExistsFunc = (path) => { return true; };
```
Nit: the lambda can simply by `path => true` or since you aren't using path, it may be simpler as `_ => true`.

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,80 @@
+        [Fact] void ReadIOError()
+        {
+            SourceContext.s_fileExistsFunc = (path) => { return true; };
+            SourceContext.s_fileReadAllTextFunc = (path) => { throw new IOException(); };
```
Similarly, this can be `_ => throw new IOException()`

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,80 @@
+            Assert.Equal(SourceContext.AppSourceContext?.Git?.RevisionId, TestRevisionId);
+        }
+
+        private string ReadSampleSourceContext(string path)
```
Nit: I'd probably make this a lambda since it's only used in one test (unless you plan on using it in other tests). Either that or a local function.

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common/Source/V1/SourceContext.File.cs:
```diff
@@ -30,6 +30,16 @@ public sealed partial class SourceContext
         private static Lazy<SourceContext> s_sourceContext = new Lazy<SourceContext>(OpenParseFile);
```
It may be better now to not initialize these here and instead call `ResetAppSourceContext()` in the constructor so the initialization logic only exists in a single place.

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            SourceContext.s_fileReadAllTextFunc = ReadSampleSourceContext;
+            SourceContext.s_fileExistsFunc = (path) => { return true; };
+            Assert.Equal(SourceContext.AppSourceContext?.Git?.RevisionId, TestRevisionId);
```
Verify the `Url` as well?

Deren-Liao commented:
PTAL.

evildour commented on apis/Google.Cloud.DevTools.Common/Google.Cloud.DevTools.Common.Tests/SourceContextTest.cs:
```diff
@@ -0,0 +1,100 @@
+        [Fact]
+        public void ReadIOError()
+        {
+            SourceContext.s_fileExistsFunc = _ => { return true; };
```
Nit: I'd still remove the braces and `return` from single expression lambdas like this, but it's up to you.

Deren-Liao replied:
Missed it.  Updated.

evildour commented:
Did you mean to update the version number of Google.Cloud.DevTools.Common with this PR? It seems to have happened after my initial comments.

Deren-Liao commented:
Yes, I later updated the version number too while I updated the apis.json.
Please refer to
https://github.com/GoogleCloudPlatform/google-cloud-dotnet/pull/1269/files
.
That PR was pending on this change (adding tests).  I think his change
missed the version change at apis.json.  Since that change was simply
update the version number, I just incorporated it here.


On Tue, Jul 25, 2017 at 3:12 PM, Mike Dour <notifications@github.com> wrote:

> Did you mean to update the version number of Google.Cloud.DevTools.Common
> with this PR? It seems to have happened after my initial comments.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-dotnet/pull/1274#issuecomment-317888510>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVxZpMhaxwJvvP7iUiZjQ80dhELw5u1tks5sRmhUgaJpZM4OiK50>
> .
>

evildour commented:
Ok, LGTM

Deren-Liao commented:
Thanks Mike for doing the code review.
